lib/builder.js

```diff
@@ -0,0 +1,475 @@
+		queryBool = query.query.filtered.query.bool;
+
+		// keyword search
+		if (filters && filters.keyword) {
+			filter = {
+				'query_string' : {
+					query : filters.keyword
+				}
+			};
+
+			queryBool.must = (queryBool.must || []).concat([filter]);
+		}
+
+		// add analytics filters to the query
+		if (filters && filters.diagnostics && diagnosticsFilterKey) {
+			if (!Array.isArray(filters.diagnostics)) {
+				if (/\,/g.test(filters.diagnostics)) {
+					filters.diagnostics = filters.diagnostics.split(/\,/);
+				} else {
+					filters.diagnostics = [filters.diagnostics];
+				}
+			}
+
+			filters.diagnostics.forEach(function (stat) {
+				if (typeof diagnosticsFilterKey[stat] !== 'undefined') {
+					filterBool.must = (filterBool.must || []).concat(
+						diagnosticsFilterKey[stat]());
+				}
+			});
+		}
+
+		// add mandatory filters to the query
+		if (filters && filters.mandatory) {
+			// notEqual queries need to be processed differently than normal
+			// match queries
+			if (filters.mandatory.ne || filters.mandatory.notEqual || filters.mandatory.notEqualTo) {
+				mustNot = {
+					'ne': filters.mandatory.ne || filters.mandatory.notEqual || filters.mandatory.notEqualTo
+				}
+
+				// remove notEqual parameters so they aren't processed again
+				filters.mandatory.ne = undefined;
+				filters.mandatory.notEqual = undefined;
+				filters.mandatory.notEqualTo = undefined;
+
+				buildFilters(mustNot).forEach(function (filter) {
+					// right now, only exact match is support
+					if (filter.match) {
+						queryBool['must_not'] = (queryBool['must_not'] || []).concat(filter);
+						return;
+					}
+				});
+			}
+
+
+			buildFilters(filters.mandatory).forEach(function (filter) {
+				if (filter.match) {
+					queryBool.must = (queryBool.must || []).concat(filter);
+
+					return;
+				}
+
+				filterBool.must = (filterBool.must || []).concat(filter);
+			});
+		}
+
+		// add optional filters to the query
+		if (filters && filters.optional) {
+			buildFilters(filters.optional).forEach(function (filter) {
+				if (filter.match) {
+					queryBool.should = (Array.isArray(queryBool.should) ? queryBool.should : [])
+						.concat(filter);
+
+					return;
+				}
+
+				filterBool.should = (Array.isArray(filterBool.should) ? filterBool.should : [])
+					.concat(filter);
+			});
+		}
+
+		// add projection
+		if (filters && filters.field) {
+			// ensure array
+			if (!Array.isArray(filters.field)) {
+				if (/\,/g.test(filters.field)) {
+					filters.field = filters.field.split(/\,/);
+				} else {
+					filters.field = [filters.field];
+				}
+			}
+
+			query._source = (query._source || []).concat(filters.field);
+		}
+
+		// add sorting
+		// NOTE: only works with simple cases as at this stage in the software
+		// the system unable to detect order between asc and desc as originally
+		// passed via the querystring
+		if (queryOptions.sort) {
+
+			if (typeof queryOptions.sort === 'string') {
```

Comment:
An odd OCD, but I tend to twitch a little when I see `if` blocks followed by `else if` and `else` blocks. 

What if we did something along the lines of the following:

```javascript
// if queryOptions.sort is a string, convert to an array
if (typeof queryOptions.sort === 'string') {
	queryOptions.sort = queryOptions.split(/\,/g).map((field) => field.trim());
}

// handle sort options provided as an Array
if (Array.isArray(queryOptions.sort) && queryOptions.sort.length) {
	sortOptions = queryOptions.sort;
}

// when queryOptions.sort is in object format, convert appropriately
if (validators.isPOJO(queryOptions.sort) {
	// the codez in herr
}
```